sdks/java/io/contextual-text-io/src/main/java/org/apache/beam/sdk/io/contextualtextio/RecordWithMetadata.java

```diff
@@ -17,69 +17,103 @@
  *   <li>recordNum: The ordinal number of the record in its file. {@link
- *       RecordWithMetadata#getRecordNum()}
- *   <li>recordValue: The value / contents of the record {@link RecordWithMetadata#getValue()}
+ *       RecordWithMetadata#RECORD_NUM}
+ *   <li>recordValue: The value / contents of the record {@link RecordWithMetadata#VALUE}
  *   <li>rangeOffset: The starting offset of the range (split), which contained the record, when the
- *       record was read. {@link RecordWithMetadata#getRangeOffset()}
+ *       record was read. {@link RecordWithMetadata#RANGE_OFFSET}
  *   <li>recordNumInOffset: The record number relative to the Range. (line number within the range)
- *       {@link RecordWithMetadata#getRecordNumInOffset()}
+ *       {@link RecordWithMetadata#RECORD_NUM_IN_OFFSET}
  *   <li>fileName: Name of the file to which the record belongs (this is the full filename,
- *       eg:path/to/file.txt) {@link RecordWithMetadata#getFileName()}
+ *       eg:path/to/file.txt) {@link RecordWithMetadata#RESOURCE_ID}
  * </ul>
  */
-@Experimental(Experimental.Kind.SCHEMAS)
-@DefaultSchema(AutoValueSchema.class)
-@AutoValue
-public abstract class RecordWithMetadata {
-  public abstract long getRecordOffset();
-
-  public abstract long getRecordNum();
-
-  public abstract String getValue();
-
-  public abstract long getRangeOffset();
-
-  public abstract long getRecordNumInOffset();
-
-  public abstract Builder toBuilder();
-
-  public abstract String getFileName();
-
-  public static Builder newBuilder() {
-    return new AutoValue_RecordWithMetadata.Builder();
+public class RecordWithMetadata {
+
+  public static final String RECORD_OFFSET = "recordOffset";
+  public static final String RECORD_NUM = "recordNum";
+  public static final String VALUE = "value";
+  public static final String RANGE_OFFSET = "rangeOffSet";
+  public static final String RECORD_NUM_IN_OFFSET = "recordNumInOffset";
+  public static final String RESOURCE_ID = "resourceID";
+
+  public static Schema getSchema() {
+    return Schema.builder()
+        .addInt64Field(RECORD_OFFSET)
+        .addInt64Field(RECORD_NUM)
+        .addStringField(VALUE)
+        .addInt64Field(RANGE_OFFSET)
+        .addInt64Field(RECORD_NUM_IN_OFFSET)
+        .addLogicalTypeField(RESOURCE_ID, new ResourceIdRow())
+        .build();
   }
 
-  @AutoValue.Builder
-  public abstract static class Builder {
-    public abstract Builder setRecordNum(long lineNum);
-
-    public abstract Builder setRecordOffset(long recordOffset);
-
-    public abstract Builder setValue(String Value);
-
-    public abstract Builder setFileName(String fileName);
-
-    public abstract Builder setRecordNumInOffset(long recordNumInOffset);
-
-    public abstract Builder setRangeOffset(long startingOffset);
-
-    public abstract RecordWithMetadata build();
+  /** A Logical type using Row to represent the ResourceId type. */
+  public static class ResourceIdRow implements LogicalType<ResourceId, Row> {
```

Comment:
We should make this part of the core beam package instead of embedding it here so that we can share it. We can make this private for now so it doesn't leak and move it in a follow-up PR.

Reply:
@TheNeuralBit, it seems like we need a shared place to define logical types so the URNs can be well known. Please work with Reza to define this location and migrate existing well known logical types to be defined there with a specification for each one (e.g. field names, types, field name meanings, ...).

Reply:
I just sent https://github.com/apache/beam/pull/12921 which documents micros_instant, the only logical type currently supported across languages. This can be documented in the same place.